backtracking/all_combinations.py

```diff
@@ -0,0 +1,35 @@
+# -*- coding: utf-8 -*-
+
+"""
+	In this problem, we want to determine all possible combinations of k
+	numbers out of 1 ... n. We use backtracking to solve this problem.
+	Time complexity: O(C(n,k)) which is O(n choose k) = O((n!/(k! * (n - k)!)))
+"""
+
+
+def generate_all_combinations(n: int, k: int) -> [[int]]:
+    result = []
+    create_all_state(1, n, k, [], result)
+    return result
+
+
+def create_all_state(increment, total_number, level, current_list, total_list):
+    if level == 0:
+        total_list.append(current_list[:])
+        return
+
+    for i in range(increment, total_number - level + 2):
+        current_list.append(i)
+        create_all_state(i + 1, total_number, level - 1, current_list, total_list)
+        current_list.pop()
+
+
+def print_all_state(total_list):
+    for i in total_list:
+        print(*i)
+
+
+n = 4
+k = 2
+total_list = generate_all_combinations(n, k)
+print_all_state(total_list)
```

Comment:
Please put these four lines under a `if __name__ == '__main__'`:
```python
if __name__ == '__main__':
    n = 4
    k = 2
    total_list = generate_all_combinations(n, k)
    print_all_state(total_list)
```

This should not change the functionality of your script but it should give it a new superpower:
* __python3 -m doctest -v backtracking/all_combinations.py__